### ISSUE TITLE
Is it possible to create a React like framework?

### ISSUE THREAD

**Issue by ghost** [tcr/rust-webplatform#12]

**Comment by anowell** [comment]
I'm not even sure I should share this yet, because it's at the 2-nights-of-hacking level of experimenting, but I'm inclined to believe the answer is yes, and that [my repo here](https://github.com/anowell/quasar) might give you a rough idea of what that could look like.

**Comment by ghost** [comment]
Seems good. Was expecting something a bit like jsx instead of string templates maybe this would do https://maud.lambda.xyz
Still I guess rust is not suited for web development yet.

**Comment by anowell** [comment]
yeah.. I'm thinking of using trait-based templating and supporting multiple template engines.(mustache/maud/horrorshow/etc..). I'd especially love to see some JSX ([this](https://crates.io/crates/jsx) looks quite unmaintained), as it is much closer to HTML syntax, which tends to be closer to my preference.

Of course, that's putting the cart before the horse, because there are much more basic things to solve first. And yeah, in playing with webplatform for a few days, Rust has some work to go before it is suited for frontend web development, but it's a fun space to be experimenting in.

**Comment by cramertj** [comment]
I've been working on it, too: https://github.com/cramertj/domafic-rs. @anowell we should compare notes.

**Comment by anowell** [comment]
Hey @cramertj, you might have to walk me through Domafic a bit. At quick glance, it seems to be building a DOM (presumably server-side, though it looks like the extensive use of traits may make parts of it amenable for client rendering later).

I just added some high-level thoughts about Quasar to my [README](https://github.com/anowell/quasar). In a nutshell, I'm currently experimenting in the browser aiming to achieve ergonomics similar to modern JS frameworks, a very top-down process where I write the code I want, and then a battle to make it work. Half my battles so far are debugging webplatform/emscripten, while the other half seems to be thinking in types like `Rc<RefCell<Vec<Box<Fn(&mut Data) + 'doc>>>>` to achieve some of the public API ergonomics I want. I'm hoping to experiment soon with making the templating engine swappable. I also have some early thoughts around isomorphic rust (same rendering code server and client side), but first I probably need to flush out more basic questions like "how would I represent a basic SPA with Quasar?".

**Comment by ghost** [comment]
Domafic's architecture looks a lot like elm or redux, seems good and single state which makes rust's immutability very useful in this scenario.

Isomorphic rust might be a bit hard if there is a lot of dependency on javascript code executing then we would need something like a javascript and dom implementation in rust (It seems the go community has done some awesome work in this regard with gopherjs). Example SSR, fetching data needs to be executed on the server which maybe can be solved using traits and templates. And I guess instead of mounting to a specific node, on the server side we would render that template into that node.

I would say the architecture of https://github.com/yoshuawuyts/choo seems really good for a easy string template based js framework.

**Comment by cramertj** [comment]
Thanks for the interest! Domafic's still very much WIP, but the goal is to use it for both server-side templating and isomorphic web apps (using the asmjs and webasm targets).
